pkg/model-serving-controller/controller/model_serving_controller.go

```
@@ -1229,3 +1233,22 @@ func (c *ModelServingController) deleteServingGroup(ctx context.Context, mi *wor
 	}
 	return nil
 }
+
+func (c *ModelServingController) deleteHeadlessService(services []*corev1.Service) {
+	for _, svc := range services {
+		svcDeepCopy := svc.DeepCopy()
+		// remove finalizer first to avoid being blocked by deleting headless service
+		utils.RemoveHeadlessServiceProtectionFinalizer(svcDeepCopy)
+		_, err := c.kubeClientSet.CoreV1().Services(svc.Namespace).Update(context.TODO(), svcDeepCopy, metav1.UpdateOptions{})
+		if err != nil {
+			klog.Errorf("failed to update service %s: %v to remove finalizer\n", svcDeepCopy.Name, err)
```

should return the error